File: mylibrary/src/main/java/com/xiaolu/mylibrary/load/core/LoadLayout.java
package com.xiaolu.mylibrary.load.core;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;

import com.xiaolu.mylibrary.load.LoadUtil;
import com.xiaolu.mylibrary.load.callback.BaseCallBack;
import com.xiaolu.mylibrary.load.callback.SuccessCallBack;

import java.util.HashMap;
import java.util.Map;

/**
 * ================================================
 *
 * @ProjectName: MyLibraryTool
 * @Package: com.xiaolu.mylibrary.load.core
 * @ClassName: LoadLayout
 * @Description: java类作用描述
 * @Author: xiaol
 * @CreateDate: 2020/8/18 11:17
 * @UpdateUser: xiaol
 * @UpdateDate: 2020/8/18 11:17
 * @UpdateRemark: 更新说明
 * @Version: 1.0
 * ================================================
 */
public class LoadLayout extends FrameLayout {
    private final String TAG = getClass().getSimpleName();
    private Map<Class<? extends BaseCallBack>, BaseCallBack> callbacks = new HashMap<>();
    private Context context;
    private BaseCallBack.OnReloadListener onReloadListener;
    private Class<? extends BaseCallBack> preCallback;
    private Class<? extends BaseCallBack> curCallback;
    private static final int CALLBACK_CUSTOM_INDEX = 1;

    public LoadLayout(@NonNull Context context) {
        super(context);
    }

    public LoadLayout(@NonNull Context context, BaseCallBack.OnReloadListener onReloadListener) {
        this(context);
        this.context = context;
        this.onReloadListener = onReloadListener;
    }

    public void setupSuccessLayout(BaseCallBack callback) {
        addCallback(callback);
        View successView = callback.getRootView();
        successView.setVisibility(View.INVISIBLE);
        addView(successView, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT));
        curCallback = SuccessCallBack.class;
    }

    public void setupCallback(BaseCallBack callback) {
        BaseCallBack cloneCallback = callback.copy();
        cloneCallback.setCallBack(context, onReloadListener);
        addCallback(cloneCallback);
    }

    public void addCallback(BaseCallBack callback) {
        if (!callbacks.containsKey(callback.getClass())) {
            callbacks.put(callback.getClass(), callback);
        }
    }

    public void showCallback(final Class<? extends BaseCallBack> callback) {
        checkCallbackExist(callback);
        if (LoadUtil.isMainThread()) {
            showCallbackView(callback);
        } else {
            postToMainThread(callback);
        }
    }

    public Class<? extends BaseCallBack> getCurrentCallback() {
        return curCallback;
    }

    private void postToMainThread(final Class<? extends BaseCallBack> status) {
        post(new Runnable() {
            @Override
            public void run() {
                showCallbackView(status);
            }
        });
    }

    private void showCallbackView(Class<? extends BaseCallBack> status) {
        if (preCallback != null) {
            if (preCallback == status) {
                return;
            }
            callbacks.get(preCallback).onDetach();
        }
        if (getChildCount() > 1) {
            removeViewAt(CALLBACK_CUSTOM_INDEX);
        }
        for (Class key : callbacks.keySet()) {
            if (key == status) {
                SuccessCallBack successCallback = (SuccessCallBack) callbacks.get(SuccessCallBack.class);
                if (key == SuccessCallBack.class) {
                    successCallback.show();
                } else {
                    successCallback.showWithCallback(callbacks.get(key).getSuccessVisible());
                    View rootView = callbacks.get(key).getRootView();
                    addView(rootView);
                    callbacks.get(key).onAttach(context, rootView);
                }
                preCallback = status;
            }
        }
        curCallback = status;
    }

    public void setCallBack(Class<? extends BaseCallBack> callback, Transport transport) {
        if (transport == null) {
            return;
        }
        checkCallbackExist(callback);
        transport.order(context, callbacks.get(callback).obtainRootView());
    }

    private void checkCallbackExist(Class<? extends BaseCallBack> callback) {
        if (!callbacks.containsKey(callback)) {
            throw new IllegalArgumentException(String.format("The Callback (%s) is nonexistent.", callback
                    .getSimpleName()));
        }
    }
}

File: mylibraryKotlin/src/main/java/com/xiaolu/mylibrarykotlin/load/core/LoadLayout.kt
package com.xiaolu.mylibrarykotlin.load.core

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.xiaolu.mylibrarykotlin.load.LoadUtil
import com.xiaolu.mylibrarykotlin.load.callback.BaseCallBack
import com.xiaolu.mylibrarykotlin.load.callback.SuccessCallBack
import java.util.*

/**
 * ================================================
 *
 * @ProjectName: MyLibraryTool
 * @Package: com.xiaolu.mylibrary.load.core
 * @ClassName: LoadLayout
 * @Description: java类作用描述
 * @Author: xiaol
 * @CreateDate: 2020/8/18 11:17
 * @UpdateUser: xiaol
 * @UpdateDate: 2020/8/18 11:17
 * @UpdateRemark: 更新说明
 * @Version: 1.0
 * ================================================
 */
class LoadLayout(context: Context) : FrameLayout(context) {
    private val TAG = javaClass.simpleName
    private val callbacks: MutableMap<Class<out BaseCallBack>, BaseCallBack?> = HashMap()
    private var mContext: Context? = null
    private var onReloadListener: BaseCallBack.OnReloadListener? = null
    private var preCallback: Class<out BaseCallBack>? = null
    var currentCallback: Class<out BaseCallBack>? = null
        private set

    constructor(context: Context, onReloadListener: BaseCallBack.OnReloadListener?) : this(context) {
        this.mContext = context
        this.onReloadListener = onReloadListener
    }

    fun setupSuccessLayout(callback: BaseCallBack) {
        addCallback(callback)
        val successView = callback.getRootView()
        successView!!.visibility = View.INVISIBLE
        addView(successView, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT))
        currentCallback = SuccessCallBack::class.java
    }

    fun setupCallback(callback: BaseCallBack) {
        val cloneCallback = callback.copy()
        cloneCallback!!.setCallBack(context, onReloadListener)
        addCallback(cloneCallback)
    }

    fun addCallback(callback: BaseCallBack?) {
        if (!callbacks.containsKey(callback!!.javaClass)) {
            callbacks[callback.javaClass] = callback
        }
    }

    fun showCallback(callback: Class<out BaseCallBack>) {
        checkCallbackExist(callback)
        if (LoadUtil.isMainThread) {
            showCallbackView(callback)
        } else {
            postToMainThread(callback)
        }
    }

    private fun postToMainThread(status: Class<out BaseCallBack>) {
        post { showCallbackView(status) }
    }

    private fun showCallbackView(status: Class<out BaseCallBack>) {
        if (preCallback != null) {
            if (preCallback == status) {
                return
            }
            callbacks[preCallback!!]!!.onDetach()
        }
        if (childCount > 1) {
            removeViewAt(CALLBACK_CUSTOM_INDEX)
        }
        for (key in callbacks.keys) {
            if (key == status) {
                val successCallback: SuccessCallBack? = callbacks[SuccessCallBack::class.java] as SuccessCallBack?
                if (key == SuccessCallBack::class.java) {
                    successCallback?.show()
                } else {
                    successCallback?.showWithCallback(callbacks[key]!!.successVisible)
                    val rootView = callbacks[key]!!.getRootView()
                    addView(rootView)
                    callbacks[key]!!.onAttach(context, rootView)
                }
                preCallback = status
            }
        }
        currentCallback = status
    }

    fun setCallBack(callback: Class<out BaseCallBack>, transport: Transport?) {
        if (transport == null) {
            return
        }
        checkCallbackExist(callback)
        transport.order(context, callbacks[callback]!!.obtainRootView())
    }

    private fun checkCallbackExist(callback: Class<out BaseCallBack>) {
        require(callbacks.containsKey(callback)) {
            String.format("The Callback (%s) is nonexistent.", callback
                    .simpleName)
        }
    }

    companion object {
        private const val CALLBACK_CUSTOM_INDEX = 1
    }
}